zigpy/device.py

```diff
@@ -33,8 +35,13 @@ class Status(enum.IntEnum):
 
     # No initialization done
     NEW = 0
+
+    # Node descriptor has been read
+    NODE_DESC = 3
```

Comment:
What do you think about making `device.Status` an intFlag instead? We still will be able to do all these, but also could have some additional states like device left, device rejoined and/or new_device?
So when device leaves, we could just mark it as "device left" and keep it in the database, till it rejoins next time?

Reply:
If we use the upper four bits of `status` (or extend it to 16 bits?) that seems like it should work, though with functional migrations from #711 anything is possible.

My main goal with adding a new "status" value was to provide ZHA with UI progress indicators, so that it can display some more progress messages when a device is joining. Tbh I'm not sure if this is a useful addition, since in my experience either devices go through the initialization process in less than a few seconds, or they fail before it starts.

Maybe it's better to make the device's join state a new field/db column?

Reply:
Definitely not an extra field in sql. It is an i t enum, not an 8 bit int. With enum, the device can be only in one state. As a flag, could have some variations, e.g. fully initialized and newly joined or fully initialized but rejoined.

Reply:
Can we get rid of device state/status tracking entirely?

The device status will end up in the database only when a device is fully initialized and these states can be derived from what the `Device` and `Endpoint` objects expose:

```Python
class Device:
    def __init__(self):
        # What does zigpy do with non-joined devices? Do we update the `joined` state if
        # a previously-unjoined device unexpectedly sends a request? What about if a
        # device never "leaves" but just stops responding? It seems better to store a
        # `last_seen` timestamp for the device in the database, but this will cause
        # unnecessary database writes if SQLite persists to disk frequently.
        self.joined: bool = False

        self.node_desc: zdo.types.NodeDescriptor | None = None
        self.endpoints = {...}

    @property
    def _non_zdo_endpoints(self):
        return [ep for epid, ep in self.endpoints.items() if epid != 0]

    @property
    def has_node_descriptor(self) -> bool:
        return self.node_desc not in (None, DUMMY_NODE_DESCRIPTOR)

    @property
    def did_zdo_init(self) -> bool:
        # XXX: Do any devices exist that do not respond properly to `Active_EP_rsp`?
        return bool(self._non_zdo_endpoints)

    @property
    def all_endpoints_init(self) -> bool:
        return self.did_zdo_init and all(ep.state != EndpointState.NEW for ep in self._non_zdo_endpoints)

    @property
    def is_initialized(self) -> bool:
        return self.has_node_desc and self.all_endpoints_init
```

Reply:
That would work too.

Reply:
I meant to say go for it.

Reply:
Sure. Since it's going to mess with the database, I'd like to get #711 merged to make any migrations possible. Or we could just ignore the `status` column for now and merge that PR later, since that PR in its current state it has the dummy node descriptors to make upstream code temporarily work around the "missing node descriptor" problem.

Reply:
Let's have this ready for the next release without app migration and include it after the next release to give more time for testing. 
We don't need to modify db schema for this PR, we could just only save the fully initialized device and hardcore the device status when saving to maintain compatibility in case need to rollback the zigpy version.